### PR TITLE
RATIS-1475. Use the stream level SlidingWindow.client in OrderedStrea…

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

Use the stream level SlidingWindow.client in OrderedStreamAsync

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1475

### CODE REVIEW

**Comment by guohao-rosicky** [comment]
On the Stream client, it is only necessary to ensure that requests within the stream ID are ordered, which increases throughput

**Comment by guohao-rosicky** [comment]
What do you think of this change @szetszwo

File: ratis-client/src/main/java/org/apache/ratis/client/impl/OrderedStreamAsync.java [hunk 1]
```diff
@@ -103,18 +106,27 @@ public String toString() {
...
   private final DataStreamClientRpc dataStreamClientRpc;
-  private final SlidingWindow.Client<DataStreamWindowRequest, DataStreamReply> slidingWindow;
+
   private final Semaphore requestSemaphore;
   private final TimeDuration requestTimeout;
   private final TimeoutScheduler scheduler = TimeoutScheduler.getInstance();
+  private final ClientId clientId;
+
+  private final ConcurrentMap<ClientInvocationId, SlidingWindow.Client<DataStreamWindowRequest, DataStreamReply>>
+      slidingWindows = new ConcurrentHashMap<>();
```
Comment (szetszwo) [inline ratis-client/src/main/java/org/apache/ratis/client/impl/OrderedStreamAsync.java:116]: We may move slidingWindow to DataStreamOutputImpl.  Then, we don't have to create this map.

**Review by szetszwo — CHANGES_REQUESTED** [review]
@guohao-rosicky , this is a very good observation!  We may declare the slidingWindow in DataStreamOutputImpl; see https://issues.apache.org/jira/secure/attachment/13037768/568_review.patch

File: ratis-client/src/main/java/org/apache/ratis/client/impl/OrderedStreamAsync.java (hunk 1 above)
Comment (guohao-rosicky) [reply ratis-client/src/main/java/org/apache/ratis/client/impl/OrderedStreamAsync.java]: Has been modified @szetszwo

**Comment by captainzmc** [comment]
Thanks @guohao-rosicky for the contribution. This should help with performance.
In the CI test, OrderedStreamAsync has a NullPointerException. Can you fix it?

File: ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java [hunk 2]
```diff
@@ -105,12 +107,13 @@ private DataStreamOutputImpl(RaftClientRequest request) {
       this.header = request;
       final ByteBuffer buffer = ClientProtoUtils.toRaftClientRequestProtoByteBuffer(header);
       this.headerFuture = send(Type.STREAM_HEADER, buffer, buffer.remaining());
+      this.slidingWindow = new SlidingWindow.Client<>(ClientInvocationId.valueOf(clientId, header.getCallId()));
```
Comment (szetszwo) [inline ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java:110]: Oops, we need to initialize the slidingWindow before sending the header, i.e.
```
      this.header = request;
      this.slidingWindow = new SlidingWindow.Client<>(ClientInvocationId.valueOf(clientId, header.getCallId()));
      final ByteBuffer buffer = ClientProtoUtils.toRaftClientRequestProtoByteBuffer(header);
      this.headerFuture = send(Type.STREAM_HEADER, buffer, buffer.remaining());
```

Comment (guohao-rosicky) [reply ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java]: Has been modified @szetszwo @captainzmc

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Review by captainzmc — APPROVED** [review]
+1 Thanks @guohao-rosicky for the contribution

**Pull request merged**